File: samples/vision/yolo11_pose/runtime/python/yolo11pose.py
# Copyright (c) 2025 D-Robotics Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# flake8: noqa: E501
# flake8: noqa: E402

"""Provide a YOLO11 pose estimation inference wrapper and pipeline utilities.

This module defines a lightweight YOLO11-Pose runtime wrapper built on HBM runtime.
It includes configuration definitions and a complete pose estimation pipeline
(preprocess, forward, postprocess), producing bounding boxes, class IDs, scores,
keypoint coordinates and keypoint scores.

Key Features:
    - YoloV11PoseConfig dataclass for configuring model parameters.
    - YoloV11Pose class providing pre_process, forward, post_process, predict,
      and __call__ methods.
    - Anchor-free DFL box decoding and keypoint coordinate decoding.
    - Class-wise NMS with keypoints kept aligned.

Typical Usage:
    >>> from yolo11pose import YoloV11Pose, YoloV11PoseConfig
    >>> cfg = YoloV11PoseConfig(model_path="/path/to/yolo11n_pose.hbm")
    >>> model = YoloV11Pose(cfg)
    >>> boxes, scores, cls_ids, kpts_xy, kpts_score = model(img)

Notes:
    - Requires hbm_runtime to be installed in the deployment environment.
    - Input images are expected in BGR format by default.
    - The model outputs 9 tensors: 3 scales × (cls, box, kpts).
    - Keypoint scores are returned as raw logits (not sigmoid-activated);
      visualize.draw_keypoints handles the threshold comparison internally.
"""

import os
import sys
import hbm_runtime
import numpy as np
from dataclasses import dataclass, field
from typing import Optional, Dict, Tuple, List

# Add project root to sys.path so we can import utility modules.
# Source files:
#   utils/py_utils/preprocess.py
#   utils/py_utils/postprocess.py
# (Check these files for preprocessing/postprocessing implementations.)
sys.path.append(os.path.abspath("../../../../../"))
import utils.py_utils.preprocess as pre_utils
import utils.py_utils.postprocess as post_utils


@dataclass
class YoloV11PoseConfig:
    """Configuration for initializing the YoloV11Pose model.

    This dataclass stores the model path and all runtime parameters required
    for preprocessing, inference, and postprocessing in the YOLO11-Pose pipeline.

    Attributes:
        model_path: Path to the compiled YOLO11-Pose `.hbm` model.
        resize_type: Image resize mode used during preprocessing.
            - 1: Keep aspect ratio with letterbox padding.
        score_thres: Minimum confidence threshold for filtering detections.
        nms_thres: IoU threshold used for Non-Maximum Suppression.
        reg: Number of DFL regression bins per bounding-box side. Defaults to 16.
        strides: Feature map downsampling strides for each detection scale.
        anchor_sizes: Feature map grid sizes for each detection scale.
    """
    model_path: str
    resize_type: int = 1
    score_thres: float = 0.25
    nms_thres: float = 0.7
    reg: int = 16
    strides: list = field(default_factory=lambda: [8, 16, 32])
    anchor_sizes: list = field(default_factory=lambda: [80, 40, 20])


class YoloV11Pose:
    """YOLO11 pose estimation wrapper based on HB_HBMRuntime.

    This class provides a unified inference pipeline for YOLO11-Pose, including
    input preprocessing, model execution, and postprocessing steps such as
    anchor-free DFL box decoding, keypoint coordinate decoding,
    Non-Maximum Suppression (NMS), and coordinate rescaling to original image space.

    Attributes:
        model: Loaded HBM runtime model instance.
        model_name: Name of the first loaded model.
        input_names: Input tensor name list.
        output_names: Output tensor name list.
        input_shapes: Input tensor shape dictionary.
        output_quants: Output quantization parameter dictionary.
        input_h: Model input height (pixels).
        input_w: Model input width (pixels).
        weights_static: DFL bin weights for box expectation computation.
        cfg: Model configuration object.

    Notes:
        The model outputs 9 tensors: for 3 detection scales (stride 8/16/32),
        each has [cls, box, kpts] outputs.
        Box regression uses DFL (16-bin distributions per side).
        Keypoints use 17 COCO keypoints, each with (x, y, score).
    """

    def __init__(self, config: YoloV11PoseConfig):
        """Initialize the YoloV11Pose model with the given configuration.

        Args:
            config: Configuration object containing model path and all inference
                parameters. All field semantics are defined in `YoloV11PoseConfig`.
        """
        # Load model and extract metadata
        self.model = hbm_runtime.HB_HBMRuntime(config.model_path)

        self.model_name = self.model.model_names[0]
        self.input_names = self.model.input_names[self.model_name]
        self.output_names = self.model.output_names[self.model_name]
        self.input_shapes = self.model.input_shapes[self.model_name]
        self.output_quants = self.model.output_quants[self.model_name]

        # Model input resolution (H, W)
        self.input_h = self.input_shapes[self.input_names[0]][1]
        self.input_w = self.input_shapes[self.input_names[0]][2]

        # DFL bin weights: shape (1, 1, reg) for expectation computation
        self.weights_static = np.arange(config.reg, dtype=np.float32)[np.newaxis, np.newaxis, :]

        self.cfg = config

    def set_scheduling_params(self,
                              priority: Optional[int] = None,
                              bpu_cores: Optional[list] = None) -> None:
        """Configure inference scheduling parameters.

        Args:
            priority: Inference priority in the range [0, 255].
            bpu_cores: List of BPU core indices used for inference.

        Returns:
            None
        """
        kwargs = {}
        if priority is not None:
            kwargs["priority"] = {self.model_name: priority}
        if bpu_cores is not None:
            kwargs["bpu_cores"] = {self.model_name: bpu_cores}

        if kwargs:
            self.model.set_scheduling_params(**kwargs)

    def pre_process(self,
                    img: np.ndarray,
                    image_format: Optional[str] = "BGR"
                    ) -> Dict[str, Dict[str, np.ndarray]]:
        """Preprocess an input image into model-required tensor format.

        The input image is letterbox-resized and converted from BGR to NV12
        (Y and UV planes).

        Args:
            img: Input image array in BGR format.
            image_format: Input image format. Currently only `"BGR"` is supported.

        Returns:
            A nested input tensor dictionary: `{model_name: {input_name: tensor}}`.

        Raises:
            ValueError: If an unsupported image format is provided.
        """
        if image_format == "BGR":
            resize_img = pre_utils.resized_image(
                img, self.input_w, self.input_h, self.cfg.resize_type)
            y, uv = pre_utils.bgr_to_nv12_planes(resize_img)
        else:
            raise ValueError(f"Unsupported image_format: {image_format}")

        return {
            self.model_name: {
                self.input_names[0]: y,
                self.input_names[1]: uv
            }
        }

    def forward(self, input_tensor: Dict[str, Dict[str, np.ndarray]]) -> Dict[str, np.ndarray]:
        """Execute model inference.

        Args:
            input_tensor: Preprocessed input tensor dictionary produced by `pre_process()`.

        Returns:
            A dictionary containing raw output tensors returned by the runtime.
        """
        outputs = self.model.run(input_tensor)
        return outputs

    def post_process(self,
                     outputs: Dict[str, Dict[str, np.ndarray]],
                     ori_img_w: int,
                     ori_img_h: int,
                     score_thres: Optional[float] = None,
                     nms_thres: Optional[float] = None,
                     ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
        """Convert raw model outputs into pose estimation results.

        Steps:
        1) Dequantize all output tensors.
        2) For each detection scale, filter by logit threshold and decode DFL
           boxes and keypoint coordinates.
        3) Concatenate results, apply NMS.
        4) Rescale boxes and keypoints to original image dimensions.

        Args:
            outputs: Raw output tensors from inference (as returned by `forward()`).
            ori_img_w: Width of the original input image.
            ori_img_h: Height of the original input image.
            score_thres: Confidence threshold override. If `None`, uses config value.
            nms_thres: IoU threshold for NMS override. If `None`, uses config value.

        Returns:
            A tuple containing:
                - boxes: Bounding boxes with shape `(N, 4)`, format `[x1, y1, x2, y2]`.
                - scores: Confidence scores with shape `(N,)`.
                - cls_ids: Class indices with shape `(N,)`.
                - kpts_xy: Keypoint coordinates with shape `(N, 17, 2)`.
                - kpts_score: Keypoint scores (raw logits) with shape `(N, 17, 1)`.
        """
        score_thres = score_thres if score_thres is not None else self.cfg.score_thres
        nms_thres = nms_thres if nms_thres is not None else self.cfg.nms_thres

        # Inverse sigmoid threshold for raw logit filtering
        conf_thres_raw = -np.log(1.0 / score_thres - 1.0)

        # Step 1: Dequantize all outputs
        fp32_outputs = post_utils.dequantize_outputs(
            outputs[self.model_name], self.output_quants)

        all_boxes = []
        all_scores = []
        all_ids = []
        all_kpts_xy = []
        all_kpts_score = []

        # Step 2: Decode each detection head (cls + box + kpts)
        for i, (stride, anchor_size) in enumerate(zip(self.cfg.strides, self.cfg.anchor_sizes)):
            cls_key  = self.output_names[3 * i]
            box_key  = self.output_names[3 * i + 1]
            kpts_key = self.output_names[3 * i + 2]

            # Filter by raw logit threshold, decode DFL boxes
            scores, ids, valid_indices = post_utils.filter_classification(
                fp32_outputs[cls_key], conf_thres_raw)
            dbboxes = post_utils.decode_boxes(
                fp32_outputs[box_key], valid_indices,
                anchor_size, stride, self.weights_static)

            # Decode keypoints (x, y, score) for valid detections
            anchor = post_utils.gen_anchor(anchor_size)[valid_indices]
            kpts_xy, kpts_score = post_utils.decode_kpts(
                fp32_outputs[kpts_key], valid_indices,
                anchor_size, stride, anchor)

            all_boxes.append(dbboxes)
            all_scores.append(scores)
            all_ids.append(ids)
            all_kpts_xy.append(kpts_xy)
            all_kpts_score.append(kpts_score)

        # Concatenate across all scales
        boxes      = np.concatenate(all_boxes,      axis=0)
        scores     = np.concatenate(all_scores,     axis=0)
        ids        = np.concatenate(all_ids,        axis=0)
        kpts_xy    = np.concatenate(all_kpts_xy,    axis=0)
        kpts_score = np.concatenate(all_kpts_score, axis=0)

        # Step 3: NMS
        keep = post_utils.NMS(boxes, scores, ids, nms_thres)

        # Step 4: Rescale boxes to original image
        xyxy = post_utils.scale_coords_back(
            boxes[keep], ori_img_w, ori_img_h,
            self.input_w, self.input_h, self.cfg.resize_type)

        # Step 5: Rescale keypoints to original image
        scaled_kpts_xy, scaled_kpts_score = post_utils.scale_keypoints_to_original_image(
            kpts_xy[keep], kpts_score[keep],
            ori_img_w, ori_img_h,
            self.input_w, self.input_h,
            self.cfg.resize_type)

        return xyxy, scores[keep], ids[keep], scaled_kpts_xy, scaled_kpts_score

    def predict(self,
                img: np.ndarray,
                image_format: str = "BGR",
                score_thres: Optional[float] = None,
                nms_thres: Optional[float] = None,
                ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
        """Run the complete pose estimation pipeline on a single image.

        This method internally performs preprocessing, inference, and postprocessing.

        Args:
            img: Input image array in BGR format.
            image_format: Input image format. Currently supports `"BGR"`.
            score_thres: Confidence threshold override.
            nms_thres: IoU threshold override for NMS.

        Returns:
            A tuple containing:
                - boxes: Bounding boxes with shape `(N, 4)`.
                - scores: Confidence scores with shape `(N,)`.
                - cls_ids: Class indices with shape `(N,)`.
                - kpts_xy: Keypoint coordinates with shape `(N, 17, 2)`.
                - kpts_score: Keypoint scores (raw logits) with shape `(N, 17, 1)`.
        """
        ori_img_h, ori_img_w = img.shape[:2]

        # 1) Preprocess
        input_tensor = self.pre_process(img, image_format)

        # 2) Inference
        outputs = self.forward(input_tensor)

        # 3) Postprocess
        return self.post_process(outputs, ori_img_w, ori_img_h, score_thres, nms_thres)

    def __call__(self,
                 img: np.ndarray,
                 image_format: str = "BGR",
                 score_thres: Optional[float] = None,
                 nms_thres: Optional[float] = None,
                 ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
        """Callable interface for the pose estimation pipeline.

        This method is functionally equivalent to calling `predict()`.

        Args:
            img: Input image array in BGR format.
            image_format: Input image format.
            score_thres: Confidence threshold override.
            nms_thres: IoU threshold override for NMS.

        Returns:
            Same return values as `predict()`.
        """
        return self.predict(img, image_format, score_thres, nms_thres)

File: platforms/s/samples/vision/yolo11_pose/runtime/cpp/src/yolo11pose.cpp
/*
 * Copyright (c) 2025 D-Robotics Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @file yolo11pose.cpp
 * @brief Implement the YOLO11-Pose inference pipeline using HB-DNN / UCP runtime APIs.
 *
 * This file contains the complete implementation of the YOLO11-Pose wrapper:
 * - Model loading, tensor allocation and deallocation.
 * - Letterbox resize + BGR->NV12 preprocessing.
 * - Synchronous BPU inference via hbDNNInferV2() and UCP APIs.
 * - Per-head DFL box decoding and keypoint extraction with logit-domain filtering.
 * - Class-wise NMS with keypoints kept in sync.
 * - Coordinate rescaling to original image space (inverse letterbox).
 *
 * OpenMP is used to parallelize the per-head decode step for improved throughput.
 *
 * @see yolo11pose.hpp
 */

#include "yolo11pose.hpp"
#include <omp.h>

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/**
 * @brief Filter class logits, decode DFL boxes, and decode keypoints for one detection head.
 *
 * For each spatial position (h, w):
 * 1) Read single-channel confidence logit; skip if below @p conf_thres_raw.
 * 2) Decode 4×reg DFL logits from @p bbox_tensor to (x1,y1,x2,y2) in input scale.
 * 3) Decode 17 keypoints (dx, dy, score) from @p kpts_tensor and map to input scale.
 *
 * Parallelized across H×W with OpenMP; merges thread-local buffers at the end.
 *
 * @param[in]  cls_tensor      Classification float tensor (1, H, W, 1).
 * @param[in]  bbox_tensor     Box distribution float tensor (1, H, W, 4*reg).
 * @param[in]  kpts_tensor     Keypoints float tensor (1, H, W, 17*3).
 * @param[in]  conf_thres_raw  Logit-domain confidence threshold.
 * @param[in]  grid_size       Feature map spatial size for this head.
 * @param[in]  stride          Downsampling stride for this head.
 * @param[in]  reg             Number of DFL bins per bbox side.
 * @param[out] detections      Decoded detections appended here.
 * @param[out] all_kpts        Decoded keypoints aligned with @p detections.
 */
static void filter_and_decode_kpts(
    const hbDNNTensor& cls_tensor,
    const hbDNNTensor& bbox_tensor,
    const hbDNNTensor& kpts_tensor,
    float conf_thres_raw,
    int grid_size,
    int stride,
    int reg,
    std::vector<Detection>& detections,
    std::vector<std::vector<Keypoint>>& all_kpts)
{
    const hbDNNTensorShape& shape = cls_tensor.properties.validShape;
    int H = shape.dimensionSize[1];
    int W = shape.dimensionSize[2];

    const int64_t* s_cls  = cls_tensor.properties.stride;
    const int64_t* s_bbox = bbox_tensor.properties.stride;
    const int64_t* s_kpts = kpts_tensor.properties.stride;

    const uint8_t* d_cls  = reinterpret_cast<const uint8_t*>(cls_tensor.sysMem.virAddr);
    const uint8_t* d_bbox = reinterpret_cast<const uint8_t*>(bbox_tensor.sysMem.virAddr);
    const uint8_t* d_kpts = reinterpret_cast<const uint8_t*>(kpts_tensor.sysMem.virAddr);

    // Thread-local caches to reduce contention
    const int nthreads = omp_get_max_threads();
    std::vector<std::vector<Detection>>              thread_dets(nthreads);
    std::vector<std::vector<std::vector<Keypoint>>>  thread_kpts(nthreads);

    // Parallel over spatial grid
    #pragma omp parallel for collapse(2)
    for (int h = 0; h < H; ++h) {
        for (int w = 0; w < W; ++w) {
            int tid = omp_get_thread_num();
            auto& dets_local = thread_dets[tid];
            auto& kpts_local = thread_kpts[tid];

            size_t base_cls  = h * s_cls[1]  + w * s_cls[2];
            size_t base_bbox = h * s_bbox[1] + w * s_bbox[2];
            size_t base_kpts = h * s_kpts[1] + w * s_kpts[2];

            // 1) Single-channel confidence logit (pose model uses 1-class head)
            const float* ptr_cls = reinterpret_cast<const float*>(d_cls + base_cls);
            float val = *ptr_cls;
            if (val < conf_thres_raw) continue;

            Detection det{};
            det.score    = sigmoid(val);
            det.class_id = 0;  // single-class: person

            // 2) DFL box decoding: 4 sides × reg bins
            float anchor_x = 0.5f + w;
            float anchor_y = 0.5f + h;
            float ltrb[4]  = {0.f, 0.f, 0.f, 0.f};

            for (int side = 0; side < 4; ++side) {
                float bins[16] = {};
                for (int bin = 0; bin < reg; ++bin) {
                    int ch = side * reg + bin;
                    const float* ptr_bbox = reinterpret_cast<const float*>(
                        d_bbox + base_bbox + ch * s_bbox[3]);
                    bins[bin] = *ptr_bbox;
                }
                // Softmax with max-trick
                float max_bin = bins[0];
                for (int i = 1; i < reg; ++i) if (bins[i] > max_bin) max_bin = bins[i];
                float sum = 0.f;
                float probs[16] = {};
                for (int i = 0; i < reg; ++i) { probs[i] = std::exp(bins[i] - max_bin); sum += probs[i]; }
                for (int i = 0; i < reg; ++i) ltrb[side] += probs[i] * i / sum;
            }

            det.bbox[0] = (anchor_x - ltrb[0]) * stride;
            det.bbox[1] = (anchor_y - ltrb[1]) * stride;
            det.bbox[2] = (anchor_x + ltrb[2]) * stride;
            det.bbox[3] = (anchor_y + ltrb[3]) * stride;

            // 3) Keypoints decode: 17 × (dx, dy, score)
            std::vector<Keypoint> kpts_vec(17);
            for (int k = 0; k < 17; ++k) {
                int base_ch = 3 * k;
                const float* ptr_k = reinterpret_cast<const float*>(
                    d_kpts + base_kpts + base_ch * s_kpts[3]);
                // Offsets are predicted on the grid; map to input-space pixels
                kpts_vec[k].x     = (ptr_k[0] * 2.0f + (anchor_x - 0.5f)) * stride;
                kpts_vec[k].y     = (ptr_k[1] * 2.0f + (anchor_y - 0.5f)) * stride;
                kpts_vec[k].score = ptr_k[2];  // raw logit, not sigmoid
            }

            dets_local.push_back(det);
            kpts_local.push_back(std::move(kpts_vec));
        }
    }

    // Merge thread-local results
    for (int t = 0; t < nthreads; ++t) {
        detections.insert(detections.end(),
                          std::make_move_iterator(thread_dets[t].begin()),
                          std::make_move_iterator(thread_dets[t].end()));
        all_kpts.insert(all_kpts.end(),
                        std::make_move_iterator(thread_kpts[t].begin()),
                        std::make_move_iterator(thread_kpts[t].end()));
    }
}

/**
 * @brief Class-wise NMS that keeps keypoints aligned with detections.
 *
 * @param[in]  detections  All candidate detections.
 * @param[in]  kpts        Keypoints aligned with @p detections.
 * @param[in]  iou_thresh  IoU threshold for suppression.
 * @return Pair of (kept detections, kept keypoints), order-aligned.
 */
static std::pair<std::vector<Detection>, std::vector<std::vector<Keypoint>>>
nms_with_kpts(const std::vector<Detection>& detections,
              const std::vector<std::vector<Keypoint>>& kpts,
              float iou_thresh)
{
    std::vector<Detection>              kept_dets;
    std::vector<std::vector<Keypoint>>  kept_kpts;

    // Group indices by class
    std::unordered_map<int, std::vector<size_t>> class_map;
    for (size_t i = 0; i < detections.size(); ++i)
        class_map[detections[i].class_id].push_back(i);

    for (auto& [cls_id, idx_list] : class_map) {
        // Sort by score descending
        std::sort(idx_list.begin(), idx_list.end(),
                  [&](size_t a, size_t b) {
                      return detections[a].score > detections[b].score;
                  });

        std::vector<bool> suppressed(idx_list.size(), false);

        for (size_t i = 0; i < idx_list.size(); ++i) {
            if (suppressed[i]) continue;
            size_t ki = idx_list[i];
            kept_dets.push_back(detections[ki]);
            kept_kpts.push_back(kpts[ki]);

            for (size_t j = i + 1; j < idx_list.size(); ++j) {
                if (!suppressed[j] &&
                    iou(detections[ki], detections[idx_list[j]]) > iou_thresh)
                    suppressed[j] = true;
            }
        }
    }
    return {kept_dets, kept_kpts};
}

// ---------------------------------------------------------------------------
// YOLO11Pose class
// ---------------------------------------------------------------------------

/**
 * @brief Construct a YOLO11Pose instance in an uninitialized state.
 */
YOLO11Pose::YOLO11Pose()
{
    model_count_       = 0;
    packed_dnn_handle_ = nullptr;
    dnn_handle         = nullptr;
    input_count_       = 0;
    output_count_      = 0;
    input_h            = 0;
    input_w            = 0;
    inited_            = false;
}

/**
 * @brief Initialize model resources from a *.hbm model file.
 *
 * @param[in] model_path Path to the quantized *.hbm model file.
 * @retval 0        Success.
 * @retval non-zero DNN or UCP API error.
 */
int32_t YOLO11Pose::init(const char* model_path)
{
    const char** model_name_list = nullptr;

    if (inited_) {
        fprintf(stderr, "YOLO11Pose::init() called twice\n");
        return -1;
    }

    HBDNN_CHECK_SUCCESS(hbDNNInitializeFromFiles(&packed_dnn_handle_, &model_path, 1),
                        "hbDNNInitializeFromFiles failed");
    HBDNN_CHECK_SUCCESS(hbDNNGetModelNameList(&model_name_list, &model_count_, packed_dnn_handle_),
                        "hbDNNGetModelNameList failed");
    HBDNN_CHECK_SUCCESS(hbDNNGetModelHandle(&dnn_handle, packed_dnn_handle_, model_name_list[0]),
                        "hbDNNGetModelHandle failed");
    HBDNN_CHECK_SUCCESS(hbDNNGetInputCount(&input_count_, dnn_handle),
                        "hbDNNGetInputCount failed");
    HBDNN_CHECK_SUCCESS(hbDNNGetOutputCount(&output_count_, dnn_handle),
                        "hbDNNGetOutputCount failed");

    input_tensors.resize(input_count_);
    output_tensors.resize(output_count_);

    for (int i = 0; i < input_count_; i++)
        HBDNN_CHECK_SUCCESS(hbDNNGetInputTensorProperties(&input_tensors[i].properties, dnn_handle, i),
                            "hbDNNGetInputTensorProperties failed");
    for (int i = 0; i < output_count_; i++)
        HBDNN_CHECK_SUCCESS(hbDNNGetOutputTensorProperties(&output_tensors[i].properties, dnn_handle, i),
                            "hbDNNGetOutputTensorProperties failed");

    input_h = input_tensors[0].properties.validShape.dimensionSize[1];
    input_w = input_tensors[0].properties.validShape.dimensionSize[2];

    prepare_input_tensor(input_tensors);
    prepare_output_tensor(output_tensors);

    inited_ = true;
    return 0;
}

/**
 * @brief Destructor: release tensor memory and DNN model resources.
 */
YOLO11Pose::~YOLO11Pose()
{
    for (int i = 0; i < input_count_; i++)  hbUCPFree(&(input_tensors[i].sysMem));
    for (int i = 0; i < output_count_; i++) hbUCPFree(&(output_tensors[i].sysMem));
    if (packed_dnn_handle_) hbDNNRelease(packed_dnn_handle_);
}

// ---------------------------------------------------------------------------
// Free pipeline functions
// ---------------------------------------------------------------------------

/**
 * @brief Preprocess an input BGR image into NV12 model input tensors.
 *
 * @param[in,out] input_tensors Model input tensors to fill.
 * @param[in]     img           Input BGR image.
 * @param[in]     input_w       Model input width.
 * @param[in]     input_h       Model input height.
 * @param[in]     image_format  Format string (only "BGR" supported).
 * @retval 0  Success.
 * @retval -1 Unsupported format.
 */
int32_t pre_process(std::vector<hbDNNTensor>& input_tensors,
                    cv::Mat& img,
                    const int input_w, const int input_h,
                    const std::string& image_format)
{
    if (image_format != "BGR") {
        fprintf(stderr, "Unsupported image_format: %s\n", image_format.c_str());
        return -1;
    }

    cv::Mat resized_mat;
    resized_mat.create(input_h, input_w, img.type());
    letterbox_resize(img, resized_mat);

    return bgr_to_nv12_tensor(resized_mat, input_tensors, input_h, input_w);
}

/**
 * @brief Execute synchronous BPU inference on prepared input tensors.
 *
 * @param[in,out] output_tensors Output tensors filled by the runtime.
 * @param[in]     input_tensors  Prepared input tensors.
 * @param[in]     dnn_handle     DNN model handle.
 * @param[in]     sched_param    Optional UCP scheduling parameters.
 * @retval 0        Success.
 * @retval non-zero DNN or UCP API error.
 */
int32_t infer(std::vector<hbDNNTensor>& output_tensors,
              std::vector<hbDNNTensor>& input_tensors,
              const hbDNNHandle_t dnn_handle,
              hbUCPSchedParam* sched_param)
{
    hbUCPTaskHandle_t task_handle{nullptr};

    HBDNN_CHECK_SUCCESS(hbDNNInferV2(&task_handle, output_tensors.data(), input_tensors.data(), dnn_handle),
                        "hbDNNInferV2 failed");

    hbUCPSchedParam ctrl_param;
    if (sched_param) {
        ctrl_param = *sched_param;
    } else {
        HB_UCP_INITIALIZE_SCHED_PARAM(&ctrl_param);
        ctrl_param.backend = HB_UCP_BPU_CORE_ANY;
    }
    HBUCP_CHECK_SUCCESS(hbUCPSubmitTask(task_handle, &ctrl_param), "hbUCPSubmitTask failed");
    HBUCP_CHECK_SUCCESS(hbUCPWaitTaskDone(task_handle, 0), "hbUCPWaitTaskDone failed");

    for (auto& t : output_tensors)
        hbUCPMemFlush(&t.sysMem, HB_SYS_MEM_CACHE_INVALIDATE);

    HBUCP_CHECK_SUCCESS(hbUCPReleaseTask(task_handle), "hbUCPReleaseTask failed");
    return 0;
}

/**
 * @brief Postprocess YOLO11-Pose outputs into final detections and keypoints.
 *
 * @param[in]  output_tensors Raw output tensors from inference.
 * @param[in]  config         Inference configuration.
 * @param[in]  orig_img_w     Original image width.
 * @param[in]  orig_img_h     Original image height.
 * @param[in]  input_w        Model input width.
 * @param[in]  input_h        Model input height.
 * @return Pair of (detections, keypoints_per_detection) in original image coordinates.
 */
PoseResult post_process(std::vector<hbDNNTensor>& output_tensors,
                        const YOLO11PoseConfig& config,
                        int orig_img_w, int orig_img_h,
                        int input_w, int input_h)
{
    float conf_thres_raw = -std::log(1.0f / config.score_thresh - 1.0f);

    std::vector<Detection>             all_dets;
    std::vector<std::vector<Keypoint>> all_kpts;

    // Step 1 & 2: Decode each head (output layout: [cls, box, kpts] × 3 scales)
    for (size_t s = 0; s < config.strides.size(); ++s) {
        const hbDNNTensor& cls_t  = output_tensors[3 * s + 0];
        const hbDNNTensor& bbox_t = output_tensors[3 * s + 1];
        const hbDNNTensor& kpts_t = output_tensors[3 * s + 2];

        std::vector<Detection>             head_dets;
        std::vector<std::vector<Keypoint>> head_kpts;

        filter_and_decode_kpts(cls_t, bbox_t, kpts_t, conf_thres_raw,
                               config.anchor_sizes[s], config.strides[s],
                               config.reg,
                               head_dets, head_kpts);

        all_dets.insert(all_dets.end(),
                        std::make_move_iterator(head_dets.begin()),
                        std::make_move_iterator(head_dets.end()));
        all_kpts.insert(all_kpts.end(),
                        std::make_move_iterator(head_kpts.begin()),
                        std::make_move_iterator(head_kpts.end()));
    }

    // Step 3: Class-wise NMS (keeps keypoints aligned)
    auto [final_dets, final_kpts] = nms_with_kpts(all_dets, all_kpts, config.nms_thresh);

    // Step 4: Rescale boxes and keypoints to original image coordinates
    scale_letterbox_bboxes_back(final_dets, orig_img_w, orig_img_h, input_w, input_h);
    scale_keypoints_back_letterbox(final_kpts, orig_img_w, orig_img_h, input_w, input_h);

    PoseResult result;
    result.detections = std::move(final_dets);
    result.keypoints  = std::move(final_kpts);
    return result;
}
